File: tests/scores_runners/scores_run.R
scores_run <- function(...) {
    #' Run the scores function and save the plots as images
    #'
    #' @description 
    #' This script runs the scores function and saves the plot as an image.
    #' It takes command-line arguments to specify the input model and optional parameters.
    #'
    #' @usage 
    #' Rscript scores_run.R <model.json> ["Name" "Value"] ["Name" "Value"] ...
    #'   - <model.json>: Path to a JSON file containing the model structure.
    #'   - ["Name" "Value"]: Optional parameters for scores function.
    #'
    #' @details
    #' When providing a JSON file:
    #'   - The JSON should contain a structure with fields:
    #'       - lvs: array of latent variables
    #'       - loads: loadings matrix (variables x components)
    #'       - scores: scores matrix
    #'       - var: total variance
    #'       - av: centering parameters (optional)
    #'       - sc: scaling parameters (optional)
    #'
    #' Optional Parameters:
    #'   - "ObsTest": [LxM] data set with observations to be visualized
    #'   - "PlotType": "Scatter" (default) or "Bars"
    #'   - "PlotCal": TRUE/FALSE to plot calibration data (TRUE by default)
    #'   - "Title": Title for the plots. Empty by default
    #'   - "ObsLabel": Names of the observations (indices are used by default)
    #'   - "ObsClass": Groups for different visualization (a single group by default)
    #'   - "BlurIndex": Avoid blur when adding labels (1 by default)
    #'   - "Color": Color palette ("hsv", "parula", "okabeIto" or custom)
    #'
    #' @examples
    #' # Run from command line:
    #' # Rscript scores_run.R model.json "PlotType" "Bars" "PlotCal" "FALSE"

    # Load required libraries
    if (!require("jsonlite")) {
        cat("Installing jsonlite package...\n")
        install.packages("jsonlite", repos = "https://cloud.r-project.org")
        library(jsonlite)
    }

    if (!require("ggplot2")) {
        cat("Installing ggplot2 package...\n")
        install.packages("ggplot2", repos = "https://cloud.r-project.org")
        library(ggplot2)
    }

    # Get command line arguments
    args <- commandArgs(trailingOnly = TRUE)

    if (length(args) < 1) {
        stop("Usage: Rscript scores_run.R <model.json> [\"Name\" \"Value\"] [\"Name\" \"Value\"] ...")
    }

    # Json file loading
    arg_model <- args[1]
    if (!file.exists(arg_model)) {
        stop(sprintf("File not found: %s", arg_model))
    }

    # Try to load the JSON file
    tryCatch({
        model_json <- readLines(arg_model, warn = FALSE)
        model <- fromJSON(paste(model_json, collapse = ""))
        cat(sprintf("Model loaded from: %s\n", arg_model))
        
        # Print model information for debugging
        cat("Structure of model:\n")
        str(model)
        
    }, error = function(e) {
        stop(sprintf("JSON parsing failed: %s", e$message))
    })

    # Model validation
    required_fields <- c("lvs", "loads", "scores", "var")
    for (field in required_fields) {
        if (!(field %in% names(model))) {
            stop(sprintf("Missing required field: %s", field))
        }
    }

    if (!("av" %in% names(model))) {
        model$av <- numeric(0)
    }
    if (!("sc" %in% names(model))) {
        model$sc <- numeric(0)
    }

    # Parse optional parameters
    params <- list()
    if (length(args) > 1) {
        # Check if parameters are in "Name" "Value" format or name=value format
        if (length(args) >= 3 && !grepl("=", args[2])) {
            # Handle "Name" "Value" pairs
            if ((length(args) - 1) %% 2 != 0) {
                stop("Each parameter must have a corresponding value")
            }
            
            for (i in seq(2, length(args), by = 2)) {
                param_name <- args[i]
                param_value <- args[i + 1]
                
                # Convert parameter names to match R function expectations
                param_mapping <- list(
                    "ObsTest" = "obstest",
                    "PlotType" = "plottype",
                    "PlotCal" = "plotcal",
                    "Title" = "tit",
                    "ObsLabel" = "label",
                    "ObsClass" = "classes",
                    "BlurIndex" = "blur",
                    "Color" = "color"
                )
                
                if (param_name %in% names(param_mapping)) {
                    r_param_name <- param_mapping[[param_name]]
                } else {
                    r_param_name <- tolower(param_name)
                }
                
                # Convert param_value to appropriate type
                if (grepl("^[0-9]+(\\.[0-9]+)?$", param_value)) {
                    param_value <- as.numeric(param_value)
                } else if (param_value == "TRUE" || param_value == "FALSE") {
                    param_value <- as.logical(param_value)
                } else if (grepl("^\\[.*\\]$|^c\\(.*\\)$", param_value)) {
                    # Handle array parameters 
                    param_value <- eval(parse(text = param_value))
                }
                
                params[[r_param_name]] <- param_value
            }
        } else {
            # Handle name=value pairs
            for (i in 2:length(args)) {
                parts <- strsplit(args[i], "=")[[1]]
                if (length(parts) != 2) {
                    stop(sprintf("Parameter format error: %s (should be Name=Value)", args[i]))
                }
                
                param_name <- parts[1]
                param_value <- parts[2]
                
                # Convert param_value to appropriate type
                if (grepl("^[0-9]+(\\.[0-9]+)?$", param_value)) {
                    param_value <- as.numeric(param_value)
                } else if (param_value == "TRUE" || param_value == "FALSE") {
                    param_value <- as.logical(param_value)
                } else if (grepl("^\\[.*\\]$|^c\\(.*\\)$", param_value)) {
                    # Handle array parameters - simplified version
                    param_value <- eval(parse(text = param_value))
                }
                
                params[[param_name]] <- param_value
            }
        }
    }

    # Handle default values for parameters
    if (is.null(params$label)) {
        n_cal <- nrow(model$scores)
        n_test <- ifelse(!is.null(params$obstest), nrow(params$obstest), 0)
        
        if (is.null(params$plotcal) || params$plotcal) {
            params$label <- as.character(c(1:n_cal, if (n_test > 0) 1:n_test else NULL))
        } else {
            params$label <- as.character(1:n_test)
        }
    }

    # Add dependencies
    source("../R/okabeIto.R")
    source("../R/textScatter.R")
    source("../R/plotScatter.R")
    source("../R/plotVec.R")
    source("../R/scores.R")

    # Execute scores function
    fig_handles <- do.call(scores, c(list(model = model), params))

    # Save figures
    for (i in 1:length(fig_handles)) {
        if (dev.cur() > 1) dev.off()  # Close any open graphics devices
        ggsave(
            filename = sprintf("scores_%d_r.png", i),
            plot = fig_handles[[i]],
            width = 8,
            height = 6,
            dpi = 300
        )
        cat(sprintf("Saved plot to scores_%d_r.png\n", i))
    }

    cat("All plots saved successfully.\n")
}

scores_run()

File: tests/scores_runners/scores_run.m
function scores_run(varargin)
    % This script runs the scores function and saves the plot as an image.
    %
    % It takes command-line arguments to specify the input model and optional parameters:
    %
    % Usage: scores_run <model> [Name-Value Pairs]
    %   - <model>: String representation of the model structure OR path to a JSON file.
    %   - [Name-Value Pairs]: Optional parameters for scores.
    %
    % When providing a JSON file:
    %   - The JSON should contain a structure with fields:
    %       - lvs: array of latent variables
    %       - loads: loadings matrix (variables x components)
    %       - scores: scores matrix
    %       - var: total variance
    %       - av: centering parameters (optional)
    %       - sc: scaling parameters (optional)
    %
    % Optional Parameters:
    %   - 'ObsTest': [LxM] data set with observations to be visualized
    %   - 'PlotType': str
    %       'Scatter': scatterplot (by default)
    %       'Bars': bar plot
    %   - 'PlotCal': bool
    %       false: plot only test data
    %       true: plot both calibration and test (by default)
    %   - 'Title': (str) title for the plots. Empty by default
    %   - 'ObsLabel': [Kx1] name of the observations (numbers are used by default)
    %   - 'ObsClass': [Kx1] groups for different visualization (a single group by default)
    %   - 'BlurIndex': [1x1] avoid blur when adding labels (1 by default)
    %   - 'Color': Choose a color for your data.
    %       'hsv' for hsv palette
    %       'parula' for parula palette
    %       'okabeIto' for color blindness (by default for multiple classes)
    %
    % Example usage with JSON:
    %   matlab -nodisplay -r "scores_run('model.json')"
    %   octave --no-gui -q scores_run.m "model.json" "PlotType" "Bars"
    %
    % Outputs:
    %   - Saves the plot as 'scores_matlab.png' (or multiple images if multiple plots are generated)

    if isoctave()
        args = argv();
        % Attempt to install and load jsonlab if it's not already loaded
        % try
        %     pkg load jsonlab;
        % catch
        %     try
        %         fprintf('Attempting to install jsonlab package...\n');
        %         pkg install -forge jsonlab;
        %         pkg load jsonlab;
        %     catch ME
        %         if ~isempty(strfind(ME.message, 'package not found'))
        %             fprintf('Error: The package "jsonlab" was not found on Octave Forge.\n');
        %             fprintf('Please ensure you have a working internet connection and that Octave Forge is accessible.\n');
        %             fprintf('You might need to try installing it manually from the Octave command line:\n');
        %             fprintf('  pkg install -forge jsonlab\n');
        %         else
        %             error('Octave requires the jsonlab package. Error: %s', ME.message);
        %         end
        %     end
        % end
    else
        args = varargin;
    end

    if numel(args) < 1
        error('Usage: scores_run <model.json> [Name-Value Pairs]');
    end

    % Json file loading
    argModel = args{1};
    if exist(argModel, 'file') ~= 2
        error('File not found: %s', argModel);
    end

    try
        fid = fopen(argModel, 'r');
        if fid == -1, error('Cannot open file'); end
        str = fread(fid, inf, '*char')';
        fclose(fid);
        model = jsondecode(str);
        fprintf('Model loaded from: %s\n', argModel);
        disp(['Class of model: ', class(model)]); % Debugging line
        if isstruct(model)
            disp('Model is a struct.');
            disp(fieldnames(model));
        elseif iscell(model)
            disp('Model is a cell.');
            disp(size(model));
        else
            disp('Model is neither a struct nor a cell.');
        end
    catch ME
        error('JSON parsing failed: %s', ME.message);
    end

    % Model validation
    requiredFields = {'lvs', 'loads', 'scores', 'var'};
    for i = 1:length(requiredFields)
        if ~isfield(model, requiredFields{i})
            error('Missing required field: %s', requiredFields{i});
        end
    end
    if ~isfield(model, 'av'), model.av = []; end
    if ~isfield(model, 'sc'), model.sc = []; end

    num_observations = size(model.scores, 1);
    fprintf('Number of observations in scores matrix: %d\n', num_observations);

    % Parse optional parameters
    params = {};
    if length(args) > 1
        for i = 2:2:length(args)
            if i+1 > length(args)
                error('Parameter %s needs a value', args{i});
            end
            param_name = args{i};
            param_value = args{i+1};

            % Special handling for ObsLabel
            if strcmp(param_name, 'ObsLabel')
                % Convert string representation of cell array to actual cell array
                if ischar(param_value) && startsWith(param_value, '{') && endsWith(param_value, '}')
                    try
                        % Remove braces and split by semicolons
                        clean_str = param_value(2:end-1);
                        elements = strsplit(clean_str, ';');
                        % Trim whitespace and remove quotes
                        elements = strtrim(elements);
                        elements = strrep(elements, '''', '');
                        param_value = elements(:);  % Ensure column vector
                    catch
                        warning('Could not parse ObsLabel value: %s', param_value);
                        param_value = [];
                    end
                end
            % Handle other special parameters
            elseif strcmp(param_name, 'ObsClass')
                if ischar(param_value)
                    try
                        evaluated_value = eval(param_value);
                        if ~isvector(evaluated_value)
                            warning('ObsClass value should evaluate to a vector.');
                            param_value = [];
                        else
                            param_value = evaluated_value(:); % Ensure column vector
                            if length(param_value) ~= num_observations
                                warning('Number of elements in ObsClass (%d) does not match the number of observations (%d). Setting ObsClass to empty.', length(param_value), num_observations);
                                param_value = [];
                            end
                        end
                    catch
                        warning('Could not parse ObsClass value: %s', param_value);
                        param_value = [];
                    end
                end
            elseif strcmp(param_name, 'BlurIndex')
                if ischar(param_value)
                    try
                        param_value = str2double(param_value);
                    catch
                        warning('Could not parse BlurIndex value: %s', param_value);
                        param_value = 1; % Default value
                    end
                end
            elseif strcmp(param_name, 'PlotCal') % Add this block to handle PlotCal
                if strcmp(param_value, 'true')
                    param_value = true;
                elseif strcmp(param_value, 'false')
                    param_value = false;
                else
                    warning('Could not parse PlotCal value: %s. Using default (true).', param_value);
                    param_value = true; % Default value
                end
            end

            params{end+1} = param_name;
            params{end+1} = param_value;
        end
    end

    % Execute scores function
    addpath('../matlab');
    figHandles = scores(model, params{:});

    % Save figures
    for i = 1:length(figHandles)
        figure(figHandles(i));
        saveas(gcf, sprintf('scores_%d_matlab.png', i));
    end
    close all;
end

function tf = isoctave()
    tf = exist('OCTAVE_VERSION', 'builtin') ~= 0;
end

scores_run();